contributingGuides/STYLE.md

```diff
@@ -618,6 +622,36 @@ There are several ways to use and declare refs and we prefer the [callback metho
 
 We love React and learning about all the new features that are regularly being added to the API. However, we try to keep our organization's usage of React limited to the most stable set of features that React offers. We do this mainly for **consistency** and so our engineers don't have to spend extra time trying to figure out how everything is working. That said, if you aren't sure if we have adopted something please ask us first.
 
+# React Hooks: Frequently Asked Questions
+
+## Are Hooks a Replacement for HOCs or Render Props?
+
+In most cases, a custom hook is a better pattern to use than an HOC or Render Prop. They are easier to create, understand, use and document. However, there might still be a case for a HOC e.g. if you have a component that abstracts some conditional rendering logic.
+
+## Where should I put my inline functions? I'm getting a lint error related to `jsx-no-bind`...
+
+If your inline function does not have any dependencies (i.e. props or state that it depends on) it can be removed from the component and moved to the top of the file. If it does have dependencies, then we should wrap it in `useCallback()` and pass the dependencies as an argument. At one time, we questioned whether there is some performance penalty to using `useCallback()` as a pre-optimization and the conclusion was that there is generally a higher potential cost to not using it vs. using it.
```

Comment:
@marcaaron What about using `useCallback` excessively for functions with low computational cost? While this approach can improve performance by reducing unnecessary re-renders, it's important to consider the tradeoffs involved.

Overusing `useCallback` can add unnecessary complexity to code, making it harder to understand and maintain. Additionally, memoizing functions that are cheap to compute may not provide a significant performance boost and can negatively impact performance by increasing memory usage and slowing down garbage collection.

It's important to balance optimization and readability when using `useCallback`. In general, it's best to prioritize clean, easy-to-understand code and only apply memoization techniques when it's clear that they will provide a tangible performance benefit.

Reply:
My concern comes from this use case while reviewing PR https://github.com/Expensify/App/pull/18424#discussion_r1185979071

Reply:
It's a solid argument and very well put. I really appreciate you sharing your thoughts on this as I likely do not have as much experience here. So, it's good to hear another perspective.

I think our struggle has always been in trying to achieve some consistent guidance to our team on patterns to use and defensively provide some default to fall back on as we worry that engineers can trip over some of these more complex ideas. So we tend to issue more blunt advice (easy to communicate how to code something) at the expense of some performance (even if it's not the best possible way to do it).

But maybe it's not a reasonable expectation to have that we can enforce everyone to do things **one way**. And much like the question of when to use `PureComponent` in the past the only "right" solution is that you just _have to_ do some reasoning about why something should be used or not used.

At the end of the day, it would be "correct" to advocate for people to consider the tradeoffs every time. We basically have concluded that it would be _way worse_ for people to never use these tools than to overuse them. But if this view is wrong then we should change it!

If you are interested, let's pick the conversation up [in the original Slack thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1680096510744619) since it's not too late to do something about this. And thanks so much for bringing it up.

Reply:
Great to hear that my perspective was helpful! It's definitely important to balance consistency and simplicity. I understand that you want to provide clear guidance to your team. Still, it's also important to encourage them to think critically about the tradeoffs involved in different coding patterns and make informed decisions.

I agree that it's only sometimes possible or desirable to enforce a single coding pattern across your entire team since different engineers may have different preferences or experiences. Ultimately, the best approach is to encourage communication and collaboration and provide guidance and resources to help your team make informed decisions.

Can we add an exception for using callbacks if the reviewer suggests improvement that doesn't follow this guideline but follows best practices or performance improvement? In this case, the reviewer could request adding a comment explaining its reasoning so that it's clear to anyone reading the code. What do you think about this?

Reply:
I think it's a good suggestion. This is one place where we might try to have "hard and fast" rules, but if there is some better reason to use (or in this case not to use) something then a comment to explain why feels appropriate. I want to ping @tgolen on this one to get a second opinion.

Reply:
Thanks for the ping! I actually come down more on the side of React's official guidance about when to have `useCallback` and [their stance](https://react.dev/reference/react/useCallback#should-you-add-usecallback-everywhere) is that you should just about _NEVER_ use it. I think that's pretty close to a hard-and-fast rule where we avoid it except for very specific instances (this is kind of like our prop spreading ESLint rule).

I think if we adopt a rule of "always use a lot of useCallback()s" then we will be opening ourselves up to a lot of "well, I just didn't think about if it was needed" and it will get overused and become very unreadable (I think it's pretty easy to argue that having `useCallback()` is less readable and maintainable than just a normal function definition.

I would rather push for normal function definitions that need to purposefully need to be kept lean and mean (highly performant) and that is a good practice that we want to promote.

I also like the suggestion of requiring comments that explain why you're not doing the best practice (sometimes there are good reasons for it), but other times, I think it can just boil down to lazy programming (ie. "I added useCallback() because I didn't spend the time to think through alternate ways of doing it").

Reply:
Ok so if I'm interpreting this correctly - the suggestion should be something like 

>Follow the guidance here: https://react.dev/reference/react/useCallback#should-you-add-usecallback-everywhere and when in doubt don't use `useCallback()` or `useMemo()`. If you do use them you should leave a comment explaining why they were chosen.

I'd think maybe we could enforce that comment thing with a lint rule in the future...

Would you guys mind if we took this back to the original Slack thread to get a consensus? We kind of landed on an opposite conclusion and it would be good to get those participants on the same page.

Reply:
Great reference about 'useCallback' @tgolen. I suggest adding [that](https://react.dev/reference/react/useCallback#should-you-add-usecallback-everywhere) to the guidelines

Reply:
@marcaaron Hey, I agree with most of the comments, but I think we shouldn't add a comment every time we use `useMemo` and `useCallback`, I think we try to remove the overuse of `useCallback` to make the code more readable (among other stuff) and adding a lot of comments will pollute the code. I would add comments on cases that aren't obvious the use of `useMemo` or `useCallback`. I know this doesn't gives a standard. In our internal chat @burczu propose to add in the checklist an extra item to pay attention to this.

Reply:
@gedu thanks! Let's continue the discussion in the Slack thread?